README.md
# Faraday::Logging::ColorFormatter

[![Gem Version](https://badge.fury.io/rb/faraday-logging-color_formatter.svg)](https://badge.fury.io/rb/faraday-logging-color_formatter)

Add some color to your Faraday logging output.

Before...

![before](https://user-images.githubusercontent.com/3071529/212807973-e87213da-d687-4db2-af70-5bb67eb81e0a.png)

After...

![after](https://user-images.githubusercontent.com/3071529/212807978-609ee381-f445-4d98-9994-31f7a4e5caa5.png)

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add faraday-logging-color_formatter

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install faraday-logging-color_formatter

## Usage

Add the following to your ruby program.

    require 'faraday/logging/color_formatter'

And then add the `Faraday::Logging::ColorFormatter` formatter to your logger middleware config.

    connection = Faraday.new(url: 'http://httpbingo.org') do |conn|
      conn.response :logger, ActiveSupport::Logger.new($stdout), formatter: Faraday::Logging::ColorFormatter
    end

If you want to replace the default `HTTP Request` and `HTTP Response` prefixes or adjust the indentation, pass in a `prefix` hash with your own values.

    connection = Faraday.new(url: 'http://httpbingo.org') do |conn|
      conn.response :logger, ActiveSupport::Logger.new($stdout), formatter: Faraday::Logging::ColorFormatter, prefix: { request: 'request', response: 'response', indent: 2 }
    end

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests and `rake rubocop` to run the linters. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/kobusjoubert/faraday-logging-color_formatter.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Change Log

## 0.2.0
#### 2023-01-19
* Allow to change log message indentation and request response messages using a new `prefix` option to the logger middleware.

## 0.1.1
#### 2023-01-17
* Gemspec homepage URI update to point to RubyGems.
* Gemspec changelog URI update.

## 0.1.0
#### 2023-01-11
* Initial commit.


spec/spec_helper.rb
# frozen_string_literal: true

require 'faraday/logging/color_formatter'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/faraday/logging/color_formatter_spec.rb
# frozen_string_literal: true

RSpec.describe Faraday::Logging::ColorFormatter do
  let(:log)    { StringIO.new }
  let(:logger) { Logger.new(log) }

  let(:stubs) do
    Faraday::Adapter::Test::Stubs.new do |stub|
      stub.get('/ebi')      { [200, { 'Content-Type' => 'text/plain' }, 'shrimp'] }
      stub.get('/tamago')   { [200, { 'Content-Type' => 'text/plain' }, 'egg'] }
      stub.post('/kohada')  { [200, { 'Content-Type' => 'application/json' }, '{"description":"spotted gizzard shad"}'] }
      stub.patch('/anago')  { [200, { 'Content-Type' => 'application/json' }, '{"desciption":"saltwater eel"}'] }
      stub.put('/anago')    { [200, { 'Content-Type' => 'application/json' }, '{"desciption":"saltwater eel"}'] }
      stub.delete('/ikura') { [200, { 'Content-Type' => 'text/plain' }, 'salmon roe'] }
      stub.get('/goma')     { [100, { 'Content-Type' => 'text/plain' }, 'sesame seeds'] }
      stub.get('/amaebi')   { [200, { 'Content-Type' => 'text/plain' }, 'sweet shrimp'] }
      stub.get('/kani')     { [300, { 'Content-Type' => 'text/plain' }, 'cooked crab meat'] }
      stub.get('/sake')     { [400, { 'Content-Type' => 'text/plain' }, 'salmon'] }
      stub.get('/kaiso')    { [500, { 'Content-Type' => 'text/plain' }, 'seaweed'] }
    end
  end

  let(:connection) { new_connection }

  it 'has a version number' do
    expect(described_class::VERSION).not_to be_nil
  end

  it 'new returns an insance of Faraday::Logging::ColorFormatter::Formatter' do
    expect(described_class.new(logger: logger, options: {})).to be_a(Faraday::Logging::ColorFormatter::Formatter)
  end

  it 'includes the request term' do
    connection.get('/ebi')
    log.rewind
    expect(log.read).to include("INFO -- : \e[34mHTTP Request  \e[0m")
  end

  it 'includes the response term' do
    connection.get('/ebi')
    log.rewind
    expect(log.read).to include("INFO -- : \e[34mHTTP Response  \e[0m")
  end

  it 'indents the request message' do
    new_connection(prefix: { indent: 2 }).get('/ebi')
    connection.get('/ebi')
    log.rewind
    expect(log.read).to include("INFO -- :   \e[34mHTTP Request  \e[0m")
  end

  it 'indents the response message' do
    new_connection(prefix: { indent: 2 }).get('/ebi')
    connection.get('/ebi')
    log.rewind
    expect(log.read).to include("INFO -- :   \e[34mHTTP Response  \e[0m")
  end

  it 'changes the request prefix' do
    new_connection(prefix: { request: 'API Request' }).get('/ebi')
    connection.get('/ebi')
    log.rewind
    expect(log.read).to include("INFO -- : \e[34mAPI Request  \e[0m")
  end

  it 'changes the response prefix' do
    new_connection(prefix: { response: 'API Response' }).get('/ebi')
    log.rewind
    expect(log.read).to include("INFO -- : \e[34mAPI Response  \e[0m")
  end

  it 'formats GET requests in blue' do
    connection.get('/tamago')
    log.rewind
    expect(log.read).to include("\e[34mGET http://sushi.com/tamago\e[0m")
  end

  it 'formats POST requests in green' do
    connection.post('/kohada')
    log.rewind
    expect(log.read).to include("\e[32mPOST http://sushi.com/kohada\e[0m")
  end

  it 'formats PATCH requests in yellow' do
    connection.patch('/anago')
    log.rewind
    expect(log.read).to include("\e[33mPATCH http://sushi.com/anago\e[0m")
  end

  it 'formats PUT requests in yellow' do
    connection.put('/anago')
    log.rewind
    expect(log.read).to include("\e[33mPUT http://sushi.com/anago\e[0m")
  end

  it 'formats DELETE requests in red' do
    connection.delete('/ikura')
    log.rewind
    expect(log.read).to include("\e[31mDELETE http://sushi.com/ikura\e[0m")
  end

  it 'formats 1XX responses in blue' do
    connection.get('/goma')
    log.rewind
    expect(log.read).to include("\e[34mStatus 100\e[0m")
  end

  it 'formats 2XX responses in green' do
    connection.get('/amaebi')
    log.rewind
    expect(log.read).to include("\e[32mStatus 200\e[0m")
  end

  it 'formats 3XX responses in yellow' do
    connection.get('/kani')
    log.rewind
    expect(log.read).to include("\e[33mStatus 300\e[0m")
  end

  it 'formats 4XX responses in red' do
    connection.get('/sake')
    log.rewind
    expect(log.read).to include("\e[31mStatus 400\e[0m")
  end

  it 'formats 5XX responses in red' do
    connection.get('/kaiso')
    log.rewind
    expect(log.read).to include("\e[31mStatus 500\e[0m")
  end

  private

  def new_connection(options = {})
    Faraday.new(url: 'http://sushi.com') do |builder|
      builder.response(:logger, logger, formatter: described_class, **options)
      builder.adapter(:test, stubs)
    end
  end
end


lib/faraday/logging/color_formatter/formatter.rb
# frozen_string_literal: true

require 'faraday'

module Faraday
  module Logging
    module ColorFormatter
      class Formatter < Faraday::Logging::Formatter
        ANSI_COLORS = { red: "\e[31m", green: "\e[32m", yellow: "\e[33m", blue: "\e[34m", magenta: "\e[35m", cyan: "\e[36m", reset: "\e[0m" }.freeze

        def initialize(logger:, options:)
          @prefix = { request: 'HTTP Request', response: 'HTTP Response', indent: 0 }.merge(options.delete(:prefix) || {})

          super
        end

        def request(env)
          http_request_term = prefix[:request].nil? || prefix[:request].strip.empty? ? nil : in_color(:blue) { "#{prefix[:request]}  " }
          http_request_log  = in_color(request_log_color(env.method)) { "#{env.method.upcase} #{apply_filters(env.url.to_s)}" }

          request_log = proc { "#{' ' * prefix[:indent]}#{http_request_term}#{http_request_log}" }
          public_send(log_level, &request_log)

          log_headers(nil, env.request_headers) if log_headers?(:request)
          log_body(nil, env[:body]) if env[:body] && log_body?(:request)
        end

        def response(env)
          http_response_term = prefix[:response].nil? || prefix[:response].strip.empty? ? nil : in_color(:blue) { "#{prefix[:response]}  " }
          http_response_log  = in_color(response_log_color(env.status)) { "Status #{env.status}" }

          status = proc { "#{' ' * prefix[:indent]}#{http_response_term}#{http_response_log}" }
          public_send(log_level, &status)

          log_headers(nil, env.response_headers) if log_headers?(:response)
          log_body(nil, env[:body]) if env[:body] && log_body?(:response)
        end

        private

        attr_reader :prefix

        def in_color(color)
          "#{ANSI_COLORS[color]}#{yield}#{ANSI_COLORS[:reset]}"
        end

        def request_log_color(request_method)
          case request_method
          when :post
            :green
          when :put, :patch
            :yellow
          when :delete
            :red
          else
            :blue
          end
        end

        def response_log_color(response_status)
          case response_status
          when 100..199
            :blue
          when 200..299
            :green
          when 300..399
            :yellow
          when 400..599
            :red
          end
        end
      end
    end
  end
end
